### DISCUSSION TITLE
fetch & env vars with app router & static builds

### DISCUSSION THREAD

**Discussion by bripkens** [vercel/next.js#50759] [App Router] [answered]
Hey folks,

I am desiring to use app router with `fetch`. However, the exact API endpoint is not known at build time and this is causing issues. Sample code:

```typescript
export async function UsersView() {
  const resp = await fetch(
    `${process.env.API_BASE_URL}/api/users`
  );

  // ...
}
```

The value for the env var is not known at build time – and the API is also not available at build time.

This causes issues such as the following when running `next build`:

```
> next build

- info Creating an optimized production build
- info Compiled successfully
- info Linting and checking validity of types
- info Collecting page data
[    ] - info Generating static pages (0/7)TypeError: Failed to parse URL from undefined/api/users
    at new Request (node:internal/deps/undici/undici:6949:19)
```

I can kind of circumvent the issue by disabling caching for the `fetch` request `{ cache: "no-store" }`, or by setting `export const dynamic = "force-dynamic";` on the route. However, this is not desirable. Caching, deduping etc. should all be supported, albeit dependent on the environment variable value.

**To clarify:** The definition of env vars is working fine. This question/issue is specifically about the fact that the API is not known/available at build time.

How are others handling this? 

# Update

Turned this into a Next.js issue https://github.com/vercel/next.js/issues/50828

**Comment by devjiwonchoi** [comment]
Make sure you follow the `.env` file convention of Next.js.

- `.env` // default
- `.env.development` // develop
- `.env.production` // production (build)
- `.env.local` // Overrides all .envs

> .env.local always overrides the defaults set.

Your `API_ENDPOINT` should be stored in `.env.production` for the `next build`.

```
// .env.production

// ...

API_ENDPOINT=YOUR_API_ENDPOINT

// ...
```

## Reference
- `.env` Examples: https://github.com/vercel/next.js/tree/canary/examples/environment-variables
- Environment Variables: https://nextjs.org/docs/app/building-your-application/configuring/environment-variables

**Reply by HamAndRock** [reply]
This is i think wrong answer, he is getting undefined so by adding https he would he https://undefined/.

The problem is how env vars work. If you don't call you env NEXT_PUBLIC_(name) it won't be available in static pages. But my naming it next public all occurnces of that env will be replaced with the correct value on build time and it can't be changed afterwards (it replaces process.env with the string from env, so without modifying build files you can't change them). If you want to have it dynamic, changing the env every run you have to make those pages using non next public env dynamically rendered.

This is new behavior that has been here for a while and it's confusing for a lot of people. There is a lot of discussion about it.

In your case rename the env to NEXT_PUBLIC_API_BASE_URL and supply the value into that variable nam during buildtime. If you need to change that value after build, there are some hacky workarounds with entry point in docker files, which basically before next start run she'll script replacing the hardcoded values in build times with the correct ones.

**Reply by devjiwonchoi** [reply]
You're right, seems like it's nothing to do with HTTP(S) and I'll remove that option. As far as I know, `NEXT_PUBLIC_` is used for exposing on the client side, isn't `static build` occurs on the server side? Could you explain more on "If you don't call you env NEXT_PUBLIC_(name) it won't be available in static pages."?

**Reply by HamAndRock** [reply]
Well it's for client side and server side. Since app router is by default static so all pages are pre built. Using process.env.ENV can't work because those pages are compiled during build. And since they are static they can't access process and always return undefined. If you opt a page to dynamic rendering, you can use env vars without next public prefix because that page is evaluated on server and has access to process.env.

With next public prefix your are telling next to bake buildtime vars into the js chunks by replacing process.env. so they are basically strings that can read by SSR or client.


This is my understanding of envs in current state. Maybe i am wrong, but this is my experience, i was working with .env vars a lot recently.

**Reply by devjiwonchoi** [reply]
@HamAndRock This is what I got.
<img width="1440" alt="Screenshot 2023-06-05 at 4 28 04 AM" src="https://github.com/vercel/next.js/assets/120007119/245c7b13-3bd5-493c-88f0-34dbfb87d9b7">

<img width="282" alt="Screenshot 2023-06-05 at 4 28 13 AM" src="https://github.com/vercel/next.js/assets/120007119/b42cb7d5-0bb5-49d7-bd1c-329c31fd91cd">

Seems like both of 'em are displayed on production.
Did I miss something?

**Reply by HamAndRock** [reply]
Nah it's probably me, you are right it will work in server components. I got bit confused, the main problem with this is when you need client side env that has to change on next start with env passed. For example the API URL basepath.

Sorry for the confusion

**Reply by bripkens** [reply]
To clarify: The definition of env vars is not the issue. I'll update the question description.

The issue is that the API endpoint is not known & not available at build time.

**Reply by HamAndRock** [reply]
Yep in that case you have to resolve to the hack with putting some dummy public env vars.. (NEXT_PUBLIC_API=NEXT_PUBLIC_API)

Before executing the next start command you add entry point pointing to bash script that walks through compiled JavaScript from next and replaces the string "NEXT_PUBLIC_API" with env from process that you have access to during start.

**Reply by HamAndRock** [reply]
https://github.com/vercel/next.js/discussions/17641#discussioncomment-339555

Here is your answer.

**Reply by bripkens** [reply]
Not sure I understand how this would help. Can you please elaborate?

If I do this, the `fetch` request will continue to be issued at build time, but fail due to URL structure issues:

```
- info Compiled successfully
- info Linting and checking validity of types
- info Collecting page data
[    ] - info Generating static pages (0/7)TypeError: Failed to parse URL from API_ORIGIN/api/users
    at new Request (node:internal/deps/undici/undici:6949:19)
    at r (ui/.next/server/chunks/269.js:25226:65)
    at doOriginalFetch (ui/.next/server/chunks/630.js:3399:24)
    at ui/.next/server/chunks/630.js:3517:20 {
  [cause]: TypeError [ERR_INVALID_URL]: Invalid URL
```

Update: Saw your follow-up comment after posting this, will check out the linked issue.
Update 2: The follow-up comment tries to address a separate issue unrelated to the app-router problem.

**Reply by devjiwonchoi** [reply]
@bripkens Just to make sure, are your API_ORIGIN include `https://` or `http://`?

**Reply by bripkens** [reply]
@devjiwonchoi: Yes it does. The value is however not known at build time (meaning I as the programmer don't know it) and consequently I cannot define the env var at build time in an env file.

**Reply by devjiwonchoi** [reply]
I got couple questions due to the lack of your project information.

1. What is the name of your `.env` file?
2. Where is your `UsersView` Page Component located? - Assuming it's a page component.

**Reply by HamAndRock** [reply]
So just put a dummy valid url in there, you can put there https://ENV/ so you can get around static analysis during build and then you would just replace the ENV with bash

**Reply by bripkens** [reply]
@devjiwonchoi

1. `.env.local`. Please note that env definition technically works. It is just that the API call **must not happen during build time** as the API is **only available at runtime**.
2. It is an app router component. So: `src/app/users/page.tsx`

**Reply by devjiwonchoi** [reply]
Don't you get an error if the page component is not in `exported default`?

```js
export default async function UsersView() {
  const resp = await fetch(
    `${process.env.API_BASE_URL}/api/users`
  );

  // ...
}
```

**Reply by bripkens** [reply]
@HamAndRock: This unfortunately doesn't work either, because the `fetch` call will continue to happen to prerender the route.

```> next build

- info Loaded env from ui/.env.local
- info Creating an optimized production build
- info Compiled successfully
- info Linting and checking validity of types
- info Collecting page data
[    ] - info Generating static pages (0/7)TypeError: fetch failed
    at Object.fetch (node:internal/deps/undici/undici:14152:11)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {
  cause: Error: getaddrinfo ENOTFOUND api_origin
      at GetAddrInfoReqWrap.onlookup [as oncomplete] (node:dns:107:26)
      at GetAddrInfoReqWrap.callbackTrampoline (node:internal/async_hooks:130:17) {
    errno: -3008,
    code: 'ENOTFOUND',
    syscall: 'getaddrinfo',
    hostname: 'api_origin'
  }
}

Error occurred prerendering page "/users". Read more: https://nextjs.org/docs/messages/prerender-error
TypeError: fetch failed
    at Object.fetch (node:internal/deps/undici/undici:14152:11)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
- info Generating static pages (7/7)

> Export encountered errors on following paths:
        /users/page: /users
```

Thank you for your answers thus far!

It is getting late over here. I'll create a reproducer tomorrow to make this discussion easier.

**Reply by devjiwonchoi** [reply]
Yeah, it's 5:24 AM here I probably had better answers if I had enough sleep. Apologies for my answers weren't professional. I'll look into your issue after I get some sleep. If you don't mind, could you tag me in order for me to know that you've opened another discussion?

**Reply by HamAndRock** [reply]
Oh yeah, this won't work with generateStaticProps :/ We had same issues as you did and many others have with this exact problem, but i haven't even realized that this hack won't overcome the genereteStaticProps functionality. So we solved it by changing our infra so we have just 2 static API base path.

I honestly now, I don't even know how would they solve this issue if you want to prebuild data based on url that is known only after building the app. It's just impossible without time travel :D. Well you can always drop genereteStatic props and just rely on fetch cach which now kinda solves the problem, but yeah not ideal.

**Reply by bripkens** [reply]
Here is the reproducer: https://github.com/bripkens/env-var-reproducer

> I honestly now, I don't even know how would they solve this issue if you want to prebuild data based on url that is known only after building the app. It's just impossible without time travel :D. Well you can always drop genereteStatic props and just rely on fetch cach which now kinda solves the problem, but yeah not ideal.

IMHO it would be great if there would be an option to have caching support at runtime, but skip population of the cache at build time.

https://nextjs.org/docs/app/api-reference/file-conventions/route-segment-config#dynamic

**Reply by devjiwonchoi** [reply]
Does this reproducer fail `next build` for you?

**Reply by devjiwonchoi** [reply]
Haven't modified any code, and it seems to be working.

Environment Variables:

```
// .env.local

API_ORIGIN=http://localhost:8000
```

Server:

```js
// server

import express from 'express'
import { Request, Response } from 'express'

const app = express()

app.get('/api/users', (req: Request, res: Response) => {
  res.send([
    { fullName: 'devjiwonchoi' },
    { fullName: 'bripkens' },
    { fullName: 'HamAndRock' },
  ])
})

app.listen(8000, () => {
  console.log('server started')
})
```

Build Console:

```console
dev-marln@Jeanui-MacBookAir env-var-reproducer % npm run build

> env-var-reproducer@0.1.0 build
> next build

- info Loaded env from /Users/dev-marln/SideProjects/env-var-reproducer/.env.local
- info Creating an optimized production build  
- info Compiled successfully
- info Linting and checking validity of types  
- info Collecting page data  
- info Generating static pages (5/5)
- info Finalizing page optimization  

Route (app)                                Size     First Load JS
┌ ○ /                                      4.77 kB        81.9 kB
├ ○ /favicon.ico                           0 B                0 B
└ ○ /users                                 137 B          77.3 kB
+ First Load JS shared by all              77.1 kB
  ├ chunks/2443530c-d4a6bd7fc7f21e3c.js    50.5 kB
  ├ chunks/488-b4b91b085767513b.js         24.7 kB
  ├ chunks/main-app-e4834958d5de7f5e.js    215 B
  └ chunks/webpack-13cb6a530b2d87f9.js     1.64 kB

Route (pages)                              Size     First Load JS
─ ○ /404                                   181 B          74.7 kB
+ First Load JS shared by all              74.5 kB
  ├ chunks/framework-8883d1e9be70c3da.js   45.1 kB
  ├ chunks/main-21bf16b99c8e734a.js        27.6 kB
  ├ chunks/pages/_app-b555d5e1eab47959.js  195 B
  └ chunks/webpack-13cb6a530b2d87f9.js     1.64 kB

○  (Static)  automatically rendered as static HTML (uses no initial props)
```

Output:
![Screenshot 2023-06-05 at 6 43 42 PM](https://github.com/vercel/next.js/assets/120007119/5a5e9cc9-ffec-4605-ac2b-ea92d19d1ac1)

**Reply by bripkens** [reply]
> Haven't modified any code, and it seems to be working.

You have modified the code: You have placed an `.env.local` file in it 😄.

We seem to be misunderstanding each other. Let me try to rephrase: During build time there is no API server. As a result, pre-rendering the page at build time will always run into one of two issues as I see it:

 1. The env var is undefined and consequently `undefined/api/users` is passed to `fetch`. `fetch` will then fail at build time with an `TypeError [ERR_INVALID_URL]: Invalid URL` error.
 2. We put a dummy URL into the env var (`API_ORIGIN=http://DUMMY_URL`) and consequently `http://DUMMY_URL/api/users` is passed to `fetch`. The DNS lookup will fail at build time with `Error: getaddrinfo ENOTFOUND dummy_url`.

In either case, the build step will fail with status code `1`.

**Reply by devjiwonchoi** [reply]
Why would you try to create a static page with 'fetch' with an invalid URL?

In that case, the server side fetch will never work. If the API Server is not running during the build, I'd recommend you to use Dynamic fetch instead of static.

**Reply by bripkens** [reply]
Here is the thing: I never stated that the page should be static. And to quote my original question:

> I can kind of circumvent the issue by disabling caching for the fetch request `{ cache: "no-store" }`, or by setting `export const dynamic = "force-dynamic";` on the route. However, this is not desirable. Caching, deduping etc. should all be supported, albeit dependent on the environment variable value.

**Reply by HamAndRock** [reply]
If all you need is to get over that error for fetch during build, just wrap it into try catch, that will still show an error but build will be successfull and on production the API_URL should be already known and it will be cached accordingly when user visits a page. But you wont get the benefit of prebuilding [pages].

**Reply by bripkens** [reply]
That will work, yes. I am trying to learn however whether there is a "proper" way of doing this. Currently I am getting the feeling that there is a missing element in the Next.js app router story.

**Reply by HamAndRock** [reply]
That's how we do it :/. Yeah as far as I know, this is "valid" solution in a way from my point of view. There are no other ways as far as I know.

**Comment by mxmknnv** [comment]
I have a similar case. My infrastructure is configured in a such way that my frontend and backend are building in parallel. So during the Next build phase my API server does not work - it cannot respond to requests with data. I understand that I can use requests without caching, but I want to take full advantage of static generation and caching. Is it possible to set this up somehow?

**Reply by bripkens** [reply]
Static generation at build time is only possible when the API server is available at build time.

**Reply by Rehan-Ul-Haq** [reply]
> Static generation at build time is only possible when the API server is available at build time.

I got similar kind of problem. But my API server is available at build time. Here is my scenario
I want to benefit from SSG and using `export const dynamic = 'force-static` . This route is also calling a fetch api. The `BASE_URL` is defined in environment varaible on vercel. Now the problem i am encountering is:
- During build time, i get `BASE_URL` undefined and so my deployment fails. 
- When i switch to `force-dynamic`, it builds successfully. But I want the benefit of SSG so this isn't the solution. 
- I tried with `force-static` and using env variables with `NEXT_PUBLIC` prefix, in that case, it gets env vars during build time and deploys successfully. 

Now what I have concluded, during build time, env var defined without `NEXT_PUBLIC` prefix aren't available and returns undefined. So I can't take the benefit of SSG (using fetch api) in my case. 

Please let me know if I am doing something wrong or any reasonable workaround to that.

**Comment by bripkens** [comment, marked as answer]
There is now a workaround for this that leverages Next.js' `headers` and `middleware` capability to circumvent these current limitations. For more information, see this issue comment.

https://github.com/vercel/next.js/issues/50828#issuecomment-1629375760
